arcgis-java-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/UtilityNetworkTraceTool.java

```diff
@@ -0,0 +1,1010 @@
+                                    traceResultInProgress.getResultsGraphicsOverlay().getGraphics().add(graphic);
+                                }
+                                // add the graphics to the MapView's graphics overlay
+                                getMapView().getGraphicsOverlays().add(traceResultInProgress.getResultsGraphicsOverlay());
+                            } else if (utilityTraceResult instanceof UtilityFunctionTraceResult) {
+                                // handle function results
+                                var functionTraceResult = (UtilityFunctionTraceResult) utilityTraceResult;
+                                functionTraceResult.getFunctionOutputs().forEach(functionOutput ->
+                                        traceResultInProgress.getFunctionResults().add(functionOutput));
+                            }
+                        }
+                        // if there were no utility element results, complete the feature future
+                        if (!containsUtilityElementResult) {
+                            fetchFeaturesForElementsCompletableFuture.complete(null);
+                        }
+                        // if the trace is successful, complete the trace future
+                        traceCompletableFuture.complete(null);
+                    } catch (CancellationException e) {
+                        // if the listenable future was cancelled complete both futures
+                        traceCompletableFuture.completeExceptionally(e);
+                        fetchFeaturesForElementsCompletableFuture.complete(null);
+                    } catch (Exception e) {
+                        // if trace fails due to another reason, set the error to the result and complete both futures
+                        traceResultInProgress.setException(e);
+                        traceCompletableFuture.complete(null);
+                        fetchFeaturesForElementsCompletableFuture.complete(null);
+                    }
+                });
+            } catch (CancellationException e) {
+                // if the listenable future was cancelled complete both futures
+                traceCompletableFuture.completeExceptionally(e);
+                fetchFeaturesForElementsCompletableFuture.complete(null);
+            } catch (Exception ex) {
+                // if trace fails due to another reason, set the error to the result and complete both futures
+                traceResultInProgress.setException(ex);
+                traceCompletableFuture.complete(null);
+                fetchFeaturesForElementsCompletableFuture.complete(null);
+            }
+        }
+    }
+
+    /**
+     * Handles clicks on the MapView. Identifies starting points if the adding starting points property is true and if
+     * a utility network is selected.
+     *
+     * @param e the mouse event
+     * @since 100.15.0
+     */
+    private void onMapViewClicked(MouseEvent e) {
```

Comment:
Hmmm I'm troubled by this architecture.  You've set up a control which swallows up all click events on the mapview.

If in the main app if I want to add a onMapViewClicked event nothing happens because you are not passing the event back up the stack.  

I'd get some guidance from @colinanderson about how to handle UI events.  You might need to look at this some more.

Reply:
Updated the code, see what you think. I've tried to leave it as flexible for the user as possible without removing the default functionality of the control. See Slack convo for more discussion.